storage/job_storage.go

```diff
@@ -0,0 +1,392 @@
+	if !exists {
+		return jobs, nil
+	}
+
+	var identifiers map[string]struct{}
+	err = j.db.Compressor().Decode("", v, &identifiers)
+	if err != nil {
+		return nil, fmt.Errorf("%w: unable to decode existing identifier", err)
+	}
+
+	for identifier := range identifiers {
+		v, err := j.Get(ctx, dbTx, identifier)
+		if err != nil {
+			return nil, fmt.Errorf("%w: unable to get job %s", err, identifier)
+		}
+
+		jobs = append(jobs, v)
+	}
+
+	return jobs, nil
+}
+
+// Ready returns all ready *job.Job.
+func (j *JobStorage) Ready(ctx context.Context, dbTx DatabaseTransaction) ([]*job.Job, error) {
+	return j.getAllJobs(ctx, dbTx, getJobMetadataKey(readyKey))
+}
+
+// Broadcasting returns all broadcasting *job.Job.
+func (j *JobStorage) Broadcasting(
+	ctx context.Context,
+	dbTx DatabaseTransaction,
+) ([]*job.Job, error) {
+	return j.getAllJobs(ctx, dbTx, getJobMetadataKey(broadcastingKey))
+}
+
+// Processing gets all processing *job.Job of a certain workflow.
+func (j *JobStorage) Processing(
+	ctx context.Context,
+	dbTx DatabaseTransaction,
+	workflow string,
+) ([]*job.Job, error) {
+	return j.getAllJobs(ctx, dbTx, getJobMetadataKey(getJobProcessingKey(workflow)))
+}
+
+// AllProcessing gets all processing *job.Jobs.
+func (j *JobStorage) AllProcessing(ctx context.Context) ([]*job.Job, error) {
+	dbTx := j.db.NewDatabaseTransaction(ctx, false)
+	defer dbTx.Discard(ctx)
+
+	return j.getAllJobs(ctx, dbTx, getJobMetadataKey(processingKey))
+}
+
+// Failed returns all failed *job.Job of a certain workflow.
+func (j *JobStorage) Failed(ctx context.Context, workflow string) ([]*job.Job, error) {
+	dbTx := j.db.NewDatabaseTransaction(ctx, false)
+	defer dbTx.Discard(ctx)
+
+	return j.getAllJobs(ctx, dbTx, getJobMetadataKey(getJobFailedKey(workflow)))
+}
+
+// AllFailed returns all failed *job.Jobs.
+func (j *JobStorage) AllFailed(ctx context.Context) ([]*job.Job, error) {
+	dbTx := j.db.NewDatabaseTransaction(ctx, false)
+	defer dbTx.Discard(ctx)
+
+	return j.getAllJobs(ctx, dbTx, getJobMetadataKey(failedKey))
+}
+
+// Completed gets all successfully completed *job.Job of a certain workflow.
+func (j *JobStorage) Completed(ctx context.Context, workflow string) ([]*job.Job, error) {
+	dbTx := j.db.NewDatabaseTransaction(ctx, false)
+	defer dbTx.Discard(ctx)
+
+	return j.getAllJobs(ctx, dbTx, getJobMetadataKey(getJobCompletedKey(workflow)))
+}
+
+// AllCompleted gets all successfully completed *job.Jobs.
+func (j *JobStorage) AllCompleted(ctx context.Context) ([]*job.Job, error) {
+	dbTx := j.db.NewDatabaseTransaction(ctx, false)
+	defer dbTx.Discard(ctx)
+
+	return j.getAllJobs(ctx, dbTx, getJobMetadataKey(completedKey))
+}
+
+func (j *JobStorage) getNextIdentifier(
+	ctx context.Context,
+	dbTx DatabaseTransaction,
+) (string, error) {
+	k := getJobMetadataKey("identifier")
+	exists, v, err := dbTx.Get(ctx, k)
+	if err != nil {
+		return "", fmt.Errorf("%w: unable to get job", err)
+	}
+
+	// Get existing identifier
+	var nextIdentifier int
```

Comment:
nit: this seems like it should be named `currentIdentifier` or something like that

Reply:
I use `nextIdentifier` here just because it is the next identifier we will apply. I consider `currentIdentifier` to be the last used identifier. 🤷